corgi/core/models.py

```
@@ -984,16 +985,45 @@ def latest_components(
     ) -> "ComponentQuerySet":
         """Return components from latest builds in a single stream."""
 
+        # Constrain both the grouping of package (type/namespace/name/arch), and the final filtered
+        # set by the root components of the ProductModel.
+        # This assumes either the Product has root components which are comparable using RPM
+        # schematics, or there is only a single package for the model, not multiple versions.
+        components = self.root_components()
+
         # the concept of 'latest component' is only relevant within product boundaries
         # we want to constrain by product type/ofuri which is why we pass in model_type and ofuri
         # into the get_latest_component stored proc annotation
+        # If a product stream has multiple variants, we want to return the latest package for each
+        # variant, see CORGI-602
 
-        product_prefetch = f"{model_type.lower()}s"
-        components = self.root_components().prefetch_related(product_prefetch)
-
-        latest_components_uuids = set(
-            self._latest_components_func(components, model_type, ofuri, include_inactive_streams)
-        )
+        latest_components_uuids: set[str] = set()
+        if model_type == "ProductStream":
+            stream_with_multiple_variants = (
+                ProductStream.objects.annotate(variants_count=Count("productvariants"))
+                .filter(ofuri=ofuri, variants_count__gte=2)
+                .first()
```

I see above we don't enforce that ofuri is unique across all models of a given type. Since there could be two streams with the same ofuri, I think we should add a check that this queryset had exactly 1 result (if we found a matching ofuri), or 0 results (if there was no matching ofuri, or if that stream had only 1 variant). Then we can raise an error if more than one matching stream was found.

---

I switched to using ProductStream.objects.get(ofuri=ofuri) which will raise a DoesNotExist or MultipleObjectsReturned error.